# YSA Homework

Veri Kaynağı: Heart Attack Analysis & Prediction Dataset (https://www.kaggle.com/datasets/rashikrahmanpritom/heart-attack-analysis-prediction-dataset/data)

In [1]:
import pandas as pd

df = pd.read_csv('heart.csv')

df.head()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0   63    1   3     145   233    1        0       150     0      2.3    0   
1   37    1   2     130   250    0        1       187     0      3.5    0   
2   41    0   1     130   204    0        0       172     0      1.4    2   
3   56    1   1     120   236    0        1       178     0      0.8    2   
4   57    0   0     120   354    0        1       163     1      0.6    2   

   caa  thall  output  
0    0      1       1  
1    0      2       1  
2    0      2       1  
3    0      2       1  
4    0      2       1

### Veri seti incelenmesi
- Veri setinin genel yapısını inceleyecek kod
```python	
df.info()
df.describe()
```
- Veri setindeki eksik verileri inceleyecek kod
```python
df.isnull().sum()
```
- Veri setindeki kategorik değişkenlerin sınıflarını inceleyecek kod
```python
df["column_name"].value_counts()
```

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


### About this dataset
- Age : Age of the patient

- Sex : Sex of the patient

- exang: exercise induced angina (1 = yes; 0 = no)

- ca: number of major vessels (0-3)

- cp : Chest Pain type chest pain type

    Value 1: typical angina
    Value 2: atypical angina
    Value 3: non-anginal pain
    Value 4: asymptomatic
    trtbps : resting blood pressure (in mm Hg)

- chol : cholestoral in mg/dl fetched via BMI sensor

- fbs : (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)

- rest_ecg : resting electrocardiographic results

    Value 0: normal
    Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
    Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
- thalach : maximum heart rate achieved

- target : 0= less chance of heart attack 1= more chance of heart attack

In [3]:
df.describe()

age         sex          cp      trtbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg    thalachh        exng     oldpeak         slp         caa  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

            thall      output  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [4]:
# Null değer kontrolü
df.isnull().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
dtype: int64

null değer yok 

In [6]:
# unique değerler
df["output"].nunique()

2

In [7]:
# kaç tane veri var
print(f"veri sayısı: {df.shape[0]}")

# kaç tane sınıf var
print(f"sınıf sayısı: {df['output'].nunique()}")

# sınıfların dağılımı
print(f"sınıfların dağılımı: {df['output'].value_counts()}")

# sınıfların 100 de kaçlık dilimde olduğu
print(f"sınıfların 100 de kaçlık dilimde olduğu: {df['output'].value_counts(normalize=True)*100}")

veri sayısı: 303
sınıf sayısı: 2
sınıfların dağılımı: output
1    165
0    138
Name: count, dtype: int64
sınıfların 100 de kaçlık dilimde olduğu: output
1    54.455446
0    45.544554
Name: proportion, dtype: float64


In [8]:
X = df.drop('output', axis=1)
X

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  
0      0      1  
1      0      2  
2      0      2  
3      0      2  
4      0      2  
..   ...    ...  
298    0      3  
299    0      3  
300    2      3  
301    1      3  
302    1      2  

[303 rows x 13 columns]

In [9]:
y = df['output']
y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: output, Length: 303, dtype: int64

Veri setini train ve test olarak ayırma

In [10]:
from sklearn.model_selection import train_test_split
# Veri setini train ve test olarak ayıralım
# test_size=0.2 veri setinin %20'sini test için ayırıyor
# random_state=42 veri setinin her seferinde aynı şekilde bölünmesi için kullanılır
# X_train ve y_train eğitim için, X_test ve y_test test için kullanılacak veri setleridir
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
X_train

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
132   42    1   1     120   295    0        1       162     0      0.0    2   
202   58    1   0     150   270    0        0       111     1      0.8    2   
196   46    1   2     150   231    0        1       147     0      3.6    1   
75    55    0   1     135   250    0        0       161     0      1.4    1   
176   60    1   0     117   230    1        1       160     1      1.4    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
188   50    1   2     140   233    0        1       163     0      0.6    1   
71    51    1   2      94   227    0        1       154     1      0.0    2   
106   69    1   3     160   234    1        0       131     0      0.1    1   
270   46    1   0     120   249    0        0       144     0      0.8    2   
102   63    0   1     140   195    0        1       179     0      0.0    2   

     caa  thall  
132    0      2  
202    0      3  
196    0      2  
75     0      2  
176    2      3  
..   ...    ...  
188    1      3  
71     1      3  
106    1      2  
270    0      3  
102    2      2  

[242 rows x 13 columns]

In [12]:
y_train

132    1
202    0
196    0
75     1
176    0
      ..
188    0
71     1
106    1
270    0
102    1
Name: output, Length: 242, dtype: int64

In [13]:
X_test

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
179   57    1   0     150   276    0        0       112     1      0.6    1   
228   59    1   3     170   288    0        0       159     0      0.2    1   
111   57    1   2     150   126    1        1       173     0      0.2    2   
246   56    0   0     134   409    0        0       150     1      1.9    1   
60    71    0   2     110   265    1        0       130     0      0.0    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
249   69    1   2     140   254    0        0       146     0      2.0    1   
104   50    1   2     129   196    0        1       163     0      0.0    2   
300   68    1   0     144   193    1        1       141     0      3.4    1   
193   60    1   0     145   282    0        0       142     1      2.8    1   
184   50    1   0     150   243    0        0       128     0      2.6    1   

     caa  thall  
179    1      1  
228    0      3  
111    1      3  
246    2      3  
60     1      2  
..   ...    ...  
249    3      3  
104    0      2  
300    2      3  
193    2      3  
184    0      3  

[61 rows x 13 columns]

In [14]:
y_test

179    0
228    0
111    1
246    0
60     1
      ..
249    0
104    1
300    0
193    0
184    0
Name: output, Length: 61, dtype: int64

In [15]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

In [16]:
X_train

array([[-1.35679832,  0.72250438,  0.00809909, ...,  0.95390513,
        -0.68970073, -0.50904773],
       [ 0.38508599,  0.72250438, -0.97189094, ...,  0.95390513,
        -0.68970073,  1.17848036],
       [-0.92132724,  0.72250438,  0.98808912, ..., -0.69498803,
        -0.68970073, -0.50904773],
       ...,
       [ 1.58263146,  0.72250438,  1.96807914, ..., -0.69498803,
         0.32186034, -0.50904773],
       [-0.92132724,  0.72250438, -0.97189094, ...,  0.95390513,
        -0.68970073,  1.17848036],
       [ 0.92942484, -1.38407465,  0.00809909, ...,  0.95390513,
         1.33342142, -0.50904773]])

In [17]:
X_test

array([[ 2.76218225e-01,  7.22504380e-01, -9.71890936e-01,
         1.16949120e+00,  5.53408401e-01, -3.83300706e-01,
        -1.04610909e+00, -1.70875171e+00,  1.47790748e+00,
        -3.75556294e-01, -6.94988026e-01,  3.21860343e-01,
        -2.19657581e+00],
       [ 4.93953764e-01,  7.22504380e-01,  1.96807914e+00,
         2.36038903e+00,  7.81171723e-01, -3.83300706e-01,
        -1.04610909e+00,  3.98288831e-01, -6.76632341e-01,
        -7.39094787e-01, -6.94988026e-01, -6.89700735e-01,
         1.17848036e+00],
       [ 2.76218225e-01,  7.22504380e-01,  9.88089118e-01,
         1.16949120e+00, -2.29363312e+00,  2.60891771e+00,
         8.43132697e-01,  1.02591793e+00, -6.76632341e-01,
        -7.39094787e-01,  9.53905134e-01,  3.21860343e-01,
         1.17848036e+00],
       [ 1.67350456e-01, -1.38407465e+00, -9.71890936e-01,
         2.16772932e-01,  3.07778522e+00, -3.83300706e-01,
        -1.04610909e+00, -5.18701733e-03,  1.47790748e+00,
         8.05943807e-01, -6.94988026e

# Klasik Makine Öğrenmesi

In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier

LR = LogisticRegression()
Svc = SVC()
RF = RandomForestClassifier()
DT = DecisionTreeClassifier()
KNN = KNeighborsClassifier(n_neighbors=25)

LR.fit(X_train, y_train)
Svc.fit(X_train, y_train)
RF.fit(X_train, y_train)
DT.fit(X_train, y_train)
KNN.fit(X_train, y_train)


print(f"LR accuracy: {LR.score(X_test, y_test)}")
print(f"SVC accuracy: {Svc.score(X_test, y_test)}")
print(f"RF accuracy: {RF.score(X_test, y_test)}")
print(f"DT accuracy: {DT.score(X_test, y_test)}")
print(f"KNN accuracy: {KNN.score(X_test, y_test)}")

LR accuracy: 0.8524590163934426
SVC accuracy: 0.8688524590163934
RF accuracy: 0.8688524590163934
DT accuracy: 0.8032786885245902
KNN accuracy: 0.8852459016393442


En yüksek doğruluğa sahip modeller RF (Random Forest Classifier) ve SVC (Support Vector Machine)

## Derin Öğrenme ile Sınıflandırma

In [20]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

giris_katmani = X_train.shape[1]
print(f"input_dim: {giris_katmani}")

output_dim = y_train.nunique()
print(f"output_dim: {output_dim}")

model = Sequential()
model.add(Dense(8, input_dim=giris_katmani, activation='relu'))
model.add(Dense(16, activation='relu', name='gizli_katman1'))
model.add(Dense(3, activation='softmax'))
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.summary() # modelin özeti

model.fit(X_train, y_train, epochs=10, batch_size=10, validation_data=(X_test, y_test))


input_dim: 13
output_dim: 2
Model: "sequential_1"


_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 8)                 112       
                                                                 
 gizli_katman1 (Dense)       (None, 16)                144       
                                                                 
 dense_3 (Dense)             (None, 3)                 51        
                                                                 
 dense_4 (Dense)             (None, 1)                 4         
                                                                 
Total params: 311 (1.21 KB)
Trainable params: 311 (1.21 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
Epoch 1/10
25/25 [==============================] - 2s 18ms/step - loss: 0.6584 - accuracy: 0.5744 - val_loss: 0.6390 - val_accuracy: 0.6557
Epoch 2/10
25/25 [===================

## Modelin Kaydedilmesi Ve Yüklenmesi

In [21]:
# Modelin kaydedilmesi
model.save('models/heart_attack.h5')

c:\Users\HP\Documents\sites\project1\derin_ogrenme\Lib\site-packages\keras\src\engine\training.py:3079: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [22]:
# Modelin yüklenmesi
from tensorflow.keras.models import load_model
yuklenen_model = load_model('models/heart_attack.h5')

# Modelin test edilmesi
yuklenen_model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 10ms/step - loss: 0.4973 - accuracy: 0.8033


[0.4972679316997528, 0.8032786846160889]